### PR TITLE
Feature/ele 777 create article as flash or 2pa2

### PR DESCRIPTION
**Pull request by dannelundqvist**
Still WiP, this PR is mainly meant to sync and merge changes with main.

* Added new functions for transforming repository format to/from grouped format (`toGroupedNewsDoc()`, `fromGroupedNewsDoc()`)
* Added new functions for tranforming to/from yjs format from grouped format (`toYjsNewsDoc()`, `fromYjsNewsDoc()` )
* Added `useModal()` hook to make it easier to create modal dialogs that is not dependant on the local component being mounted.
* Refactored app menu to make it easier to trigger creation modal dialogs (i.e. for now only flash editor)
* Added `useActiveAuthor()` that uses the logged in user email to find the author document/data for bylines etc
* Added `useActiveDocument()` that allows you to fetch the active views underlying document from the repository. Useful to fetch the planning in focus when creating a flash for example.
* Added `YBlock.create()` that creates a more correct data representation when creating blocks to insert into yjs structures. (`Block.Create()` does not do this correctly).

Also added the basis for creating a flash from the application menu. Will continue with that after merge.

### CODE REVIEW

File: src-srv/api/documents/[id]/index.ts [hunk 1]
```diff
@@ -1,15 +1,86 @@
...
+    const { document, version } = toGroupedNewsDoc(doc)
+
+    return {
+      payload: {
+        version,
+        document
+      }
+    }
+  } catch (ex) {
+    console.error(ex)
```
Comment (gularsson) [inline src-srv/api/documents/[id]/index.ts:79]: Use pino `logger.error(ex)`

Comment (gularsson) [reply src-srv/api/documents/[id]/index.ts:79]: Or maybe throw and let middleware handle

File: src-srv/utils/transformations/groupedNewsDoc.ts [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+  if (!document) {
+    throw new Error('GetDocumentResponse contains no document')
+  }
+
+  if (document.type === 'core/planning-item') {
+    assertPlanningHasNecessaryProperties(document)
+  }
+
+  // const { meta, links, content, ...properties } = document
```
Comment (gularsson) [inline src-srv/utils/transformations/groupedNewsDoc.ts:28]: :broom:

File: src/views/Flash/index.tsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    </>
+  )
+}
+
+const FlashViewContent = (props: ViewProps & { documentId: string }): JSX.Element | undefined => {
+  const { provider } = useCollaboration()
+  const { data, status } = useSession()
+  const [documentStatus, setDocumentStatus] = useDocumentStatus(props.documentId)
+
+  console.log(provider?.document.toJSON())
```
Comment (gularsson) [inline src/views/Flash/index.tsx:65]: :broom:

File: src/views/Flash/index.tsx [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+          <div className="flex flex-row items-center">
+            <Assignees
+              path='links.core/author'
+              name='FlashAssignees'
+              placeholder='Byline'
+            />
+            <Section />
+          </div>
+
+          {/* <AssignmentTable /> */}
```
Comment (gularsson) [inline src/views/Flash/index.tsx:130]: :broom:

File: src/components/View/ViewHeader/CreateDocumentDialog.tsx [hunk 5]
```diff
@@ -77,9 +79,9 @@ export const CreateDocumentDialog = ({ type, payload, children, mutator }: Props
 function getTemplate(type: View): (id: string) => Document {
   switch (type) {
     case 'Planning':
-      return templates.planning
+      return Templates.planning
     case 'Event':
-      return templates.event
+      return Templates.event
```
Comment (gularsson) [inline src/components/View/ViewHeader/CreateDocumentDialog.tsx:84]: What are the thoughts regarding the capital T? It's not a component, it's not a class?

File: src-srv/utils/CollaborationServer.ts [hunk 6]
```diff
@@ -322,71 +314,75 @@ export class CollaborationServer {
...
+      versionMap.set('version', result?.response.version.toString())
+      hashMap.set('hash', updatedHash)
+    }).catch(ex => {
+      throw new Error('Update document with new hash and version failed', { cause: ex })
+    })
 
-        logger.debug('::: Snapshot saved: ', result?.response.version, 'new hash:', currentHash)
-      }
-    }
+    console.debug('::: Document saved to repository: ', result.response.version, 'new hash:', updatedHash)
```
Comment (gularsson) [inline src-srv/utils/CollaborationServer.ts:383]: logger.debug?

**Review by gularsson — CHANGES_REQUESTED** [review]

File: __tests__/yMapValueByPath.test.ts
[the remark this answers is not in the document]
Comment (dannelundqvist) [reply __tests__/yMapValueByPath.test.ts:1]: It is used by yObserver and createYItem, so not yet. I will add @deprecated annotation to it.

File: __tests__/transform-document.test.ts [hunk 7]
```diff
@@ -31,83 +34,104 @@ function sortDocument(json: any): any {
...
-    if (!document || !article.document) {
-      throw new Error('no document')
-    }
+    const { documentResponse } = await fromYjsNewsDoc(yDoc)
+    const ungroupedArticle = await fromGroupedNewsDoc(documentResponse)
 
-    expect(version).toBe(article.version)
-    expect(sortDocument(document)).toEqual(sortDocument(article.document))
+    expect(ungroupedArticle.version).toBe(article.version)
+    // expect(sortDocument(ungroupedArticle.document)).toEqual(sortDocument(article.document))
```
[the remark this answers is not in the document]
Comment (dannelundqvist) [reply __tests__/transform-document.test.ts:106]: Correct!

File: src/components/App/Menu.tsx [hunk 8]
```diff
@@ -20,7 +21,8 @@ export const Menu = (): JSX.Element => {
       <SheetTrigger className="rounded-md hover:bg-gray-100 w-9 h-9 flex items-center justify-center">
         <MenuIcon strokeWidth={2.25} size={18} />
       </SheetTrigger>
-      <SheetContent side="left" className="w-100vw h-100vh p-0 flex flex-col justify-between" defaultClose={false}>
+
+      <SheetContent side="left" className="w-100vw h-100vh p-0 flex flex-col justify-between">
```
[the remark this answers is not in the document]
Comment (dannelundqvist) [reply src/components/App/Menu.tsx:25]: I'll clean it out.

File: __tests__/yMapValueByPath.test.ts
[the remark this answers is not in the document]
Comment (dannelundqvist) [reply __tests__/yMapValueByPath.test.ts:1]: I'm creating an issue in Linear as well.

File: src/components/View/ViewHeader/CreateDocumentDialog.tsx (hunk 5 above)
Comment (dannelundqvist) [reply src/components/View/ViewHeader/CreateDocumentDialog.tsx:84]: * **c**amelCase for all variables/consts and object members (properties, methods)
* **P**ascalCase for class names, constructors and static objects

Comment (dannelundqvist) [reply src/components/View/ViewHeader/CreateDocumentDialog.tsx:84]: But this is not 100%. So I can discuss it... ;)

Comment (gularsson) [reply src/components/View/ViewHeader/CreateDocumentDialog.tsx:84]: I would say UPPERCASE for constants, but I don't really like that either....

**Review by gularsson — APPROVED** [review]

**Pull request merged**